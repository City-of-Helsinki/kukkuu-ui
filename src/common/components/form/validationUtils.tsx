import { newMoment } from '../../time/utils';
import { DEFAULT_DATE_FORMAT } from '../../time/TimeConstants';

/**
 * validateRequire()
 * Check if field is required.
 * @param value
 * @param customMessage
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const validateRequire = (value: any, customMessage?: string) => {
  if (!value) {
    return customMessage || 'validation.general.required';
  }
};

/** validatePostalCode()
 * Validate postal code. We allow Finnish format: Five numbers that
 * can have leading zeroes. Example: 00100.
 * @param {string} value Input value.
 * @returns {string} translation key
 */
const validatePostalCode = (value: string): string | undefined => {
  const valid: boolean = /^\d{5}$/.test(value);
  if (value && !valid) return 'validation.postalCode.invalidFormat';
};

/** validateDate()
 * Validate input date. Dates in the future are not valid.
 * @param {string|number} value Input value.
 * @returns {string} translation key
 */
const validateDate = (value: string | number) => {
  const inputMoment = newMoment(value, DEFAULT_DATE_FORMAT);
  const nowMoment = newMoment();

  if (!inputMoment.isValid()) {
    return 'validation.date.invalidFormat';
  }

  if (inputMoment > nowMoment) {
    return 'validation.date.unSupported';
  }
};

/**
 * validateEqual()
 * Usually to check if user is located in supported area.
 * Can be used widely to check in many different case.
 * @param value Input value
 * @param comparedValue Value to compare
 * @param errorMessage Error message when not match
 */
const validateEqual = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  value: any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  comparedValue: any,
  errorMessage: string
) => {
  let cloneValue = value;
  let cloneComparedValue = comparedValue;

  if (typeof value === 'string' && typeof comparedValue === 'string') {
    cloneValue = value.toLowerCase();
    cloneComparedValue = comparedValue.toLowerCase();
  }

  if (typeof value === 'string' && Array.isArray(comparedValue)) {
    const match = comparedValue.find(c => {
      cloneValue = typeof value === 'string' ? value.toLowerCase() : value;
      cloneComparedValue = typeof c === 'string' ? c.toLowerCase() : c;

      return cloneValue === cloneComparedValue;
    });

    if (!match) return errorMessage;
  }

  if (cloneValue !== cloneComparedValue) {
    return errorMessage;
  }
};

export { validateDate, validateEqual, validatePostalCode, validateRequire };
