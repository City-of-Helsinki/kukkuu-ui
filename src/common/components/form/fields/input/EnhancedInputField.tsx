import React from 'react';
import { Field, FieldAttributes } from 'formik';

import { validateRequire } from '../../validationUtils';

interface Props {
  required: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const EnhancedInputField: React.FC<Props & FieldAttributes<any>> = ({
  required,
  validate,
  ...rest
}) => {
  let customValidate = validate;

  if (required) {
    customValidate = validateRequire || validate;
  }
  return <Field validate={customValidate} required={required} {...rest} />;
};

export default EnhancedInputField;
