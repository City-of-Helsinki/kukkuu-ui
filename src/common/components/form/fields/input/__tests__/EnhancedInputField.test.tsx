import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import EnhancedInputField from '../EnhancedInputField';
import { validateRequire } from '../../../validationUtils';

describe('EnhancedInputField ', () => {
  const wrapper = shallow(
    <EnhancedInputField name="foo" id="bar" required={true} />
  );

  test('render normally', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('validate validateRequire if value is empty and required is true', () => {
    expect(wrapper.prop('validate')).toEqual(validateRequire);
  });
});
