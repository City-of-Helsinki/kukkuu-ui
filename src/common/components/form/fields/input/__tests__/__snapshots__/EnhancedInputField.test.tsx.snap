// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnhancedInputField  render normally 1`] = `
<Field
  id="bar"
  name="foo"
  required={true}
  validate={[Function]}
/>
`;
